src/data/fintype/order.lean

```diff
@@ -0,0 +1,56 @@
+/-
+Copyright (c) 2021 Peter Nelson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Peter Nelson
+-/
+
+import order.lattice
+import data.fintype.basic
+
+/-!
+# Order structures on finite sets
+
+This file provides instances showing that a `semilattice_sup_bot` on a `fintype` gives a
+`complete_lattice`, and that a `linear_order` on a nonempty `fintype` gives a
+`complete_linear_order`. The latter applies in particular to `fin (n+1)`. The instances are
+noncomputable because the definition of `Sup` uses `set.to_finset`, which implicitly
+requires a `decidable_pred` instance for every `s : set α`.
+-/
+
+open_locale classical
+
+noncomputable def fintype.complete_semilattice_Sup (α : Type*)
+  [fintype α] [semilattice_sup_bot α] :
+  complete_semilattice_Sup α :=
+{ Sup := λ s, s.to_finset.sup id,
+  le_Sup := λ _ _ ha, finset.le_sup (set.mem_to_finset.mpr ha),
+  Sup_le := λ _ _ ha, finset.sup_le (λ b hb, ha _ (by rwa set.mem_to_finset at hb)),
+  ..(infer_instance : semilattice_sup_bot α) }
+
+noncomputable def fintype.complete_lattice_of_semilattice_sup_bot (α : Type*)
+  [fintype α] [semilattice_sup_bot α] :
+  complete_lattice α :=
+@complete_lattice_of_complete_semilattice_Sup α (fintype.complete_semilattice_Sup α)
```

Comment:
This definition feels like a bad idea to me - why not just provide `inf` explicitly as `s.to_finset.inf id?

Reply:
I don't understand what you mean here.

Reply:
I think you should add a matching `fintype.complete_semilattice_Inf` swapping `finset.sup` for `finset.inf`, then instead implement:

```suggestion
noncomputable def fintype.complete_lattice (α : Type*)
  [fintype α] [lattice α] :
  complete_lattice α :=
{ ..fintype.complete_semilattice_Sup,
  ..fintype.complete_semilattice_Inf }
```

Reply:
You need `bounded_lattice` for the second instance here, not just `lattice`. This raises the question of where to start - 
for a `fintype`,  any of `semilattice_sup_bot`, `semilattice_inf_top` and `lattice` are the same, and all give rise to `complete_lattice`. Would it make sense to have ways up to `complete_lattice` for all three?

Reply:
> You need bounded_lattice for the second instance here, not just lattice.

Good catch

> Would it make sense to have ways up to complete_lattice for all three?

`complete_lattice_of_complete_semilattice_Sup` comes with some warnings indicating that it provides bad definitional equality. I'd be inclined to avoid it entirely - if the user wants it, they can apply it to `fintype.complete_semilattice_Sup` manually.

So I would probably lean towards providing the path to `complete_lattice` only for `bounded_lattice`.

Reply:
I've just pushed a substantial rewrite which gets rid of `complete_semilattice_Sup` (/Inf) entirely, and instead just builds a bounded lattice then a complete lattice directly. This makes more of it computable.